RecoPixelVertexing/PixelTriplets/plugins/GPUCACell.h

```diff
@@ -346,4 +347,18 @@ class GPUCACell {
   hindex_type theOuterHitId;
 };
 
+template <>
+__device__ inline void GPUCACell::find_ntuplets<0>(Hits const& hh,
+                                                   GPUCACell* __restrict__ cells,
+                                                   CellTracksVector& cellTracks,
+                                                   HitContainer& foundNtuplets,
+                                                   cms::cuda::AtomicPairCounter& apc,
+                                                   Quality* __restrict__ quality,
+                                                   TmpTuple& tmpNtuplet,
+                                                   const unsigned int minHitsPerNtuplet,
+                                                   bool startAt0) const {
+  printf("ERROR: GPUCACell::find_ntuplets reached full depth!\n");
+  assert(false);  // on GPU is compiled away in NDEBUG mode
```

Comment:
My only concern would be that how could we guarantee in the future that end of the finite recursion is not reached? (in a cheap way)

I understand with the phase1 pixel detector the `6` can not be reached (or there is a bug elsewhere). The question is linked to more general https://github.com/cms-sw/cmssw/issues/32502, so perhaps this the `printf()` is ok for this PR. (even if no-one will check the printouts in production)

Reply:
What about `__trap()` (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#trap-function) ?

I've never tried, so I don't know if it works or if it has the same performance problems as `assert()`

Reply:
let's put it in a issue (including test of behaviour and performance)
in principle even if arrives here it will just create a shorter tuple (or crash somehow in the caller because the tuple is too large)
(there is another assert in the caller or in the function itself to check the size of the tuple)

Reply:
The kernels are plenty of places that would require a more robust error reporting: maybe time to ask somebody to propose a "performant framework" for that....

Reply:
Anyhow this specific "error" can not happen in production by construction. It requires major changes to the CA (such as a Phase2 setup) and in such a case it will be caught by simple tests by the developers.

Reply:
It looks like `__trap()` does not add any overhead:

measurement     | CMSSW_12_1_0_pre3 |   with #35285     |   with #35473     |  with `__trap()`  
----------------|-------------------|-------------------|-------------------|-------------------
I/O throughput  |      ~ 2 kev/s    |        ~ 2 kev/s  |        ~ 2 kev/s  |        ~ 2 kev/s  
quadruplets     |   1071 ±  3 ev/s  |   1084 ±  3 ev/s  |   1289 ±  3 ev/s  |   1292 ±  4 ev/s  
triplets        |    560 ±  2 ev/s  |    564 ±  1 ev/s  |    724 ±  3 ev/s  |    720 ±  2 ev/s  

When it triggers (simulated by decreasing the maximum from `6` to `1`), the job dies with
```
Begin processing the 1st record. Run 1, Event 1602, LumiSection 17 on stream 0 at 05-Oct-2021 15:15:12.381 CEST
ERROR: GPUCACell::find_ntuplets reached full depth!
... (repeated 819 times)
ERROR: GPUCACell::find_ntuplets reached full depth!
terminate called after throwing an instance of 'std::runtime_error'
  what():  
/data/user/fwyzard/CMSSW_12_1_0_pre3/src/HeterogeneousCore/CUDAUtilities/src/CachingDeviceAllocator.h, line 617:
cudaCheck(error = cudaEventRecord(search_key.ready_event, search_key.associated_stream));
cudaErrorLaunchFailure: unspecified launch failure
```
followed by the usual stack trace.

Reply:
I'll make a PR with
```diff
-  assert(false);  // on GPU is compiled away in NDEBUG mode
+#ifdef __CUDA_ARCH__
+  __trap();
+#else
+  abort();
+#endif
```

We may later decide if we want to wrap it in some other function.